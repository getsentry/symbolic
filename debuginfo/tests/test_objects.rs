use std::fmt;

use failure::Error;
use insta;

use symbolic_common::ByteView;
use symbolic_debuginfo::{DebugSession, Function, Object, SymbolMap};

/// Helper to create neat snapshots for symbol tables.
struct SymbolsDebug<'a>(&'a SymbolMap<'a>);

impl fmt::Debug for SymbolsDebug<'_> {
    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
        for symbol in self.0.iter() {
            writeln!(
                f,
                "{:>16x} {}",
                &symbol.address,
                &symbol.name().unwrap_or("<unknown>")
            )?;
        }

        Ok(())
    }
}

/// Helper to create neat snapshots for function trees.
struct FunctionsDebug<'a>(&'a [Function<'a>], usize);

impl fmt::Debug for FunctionsDebug<'_> {
    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
        for function in self.0 {
            writeln!(
                f,
                "\n{:indent$}> {:#x}: {} ({:#x})",
                "",
                function.address,
                function.name,
                function.size,
                indent = self.1 * 2
            )?;

            for line in &function.lines {
                writeln!(
                    f,
                    "{:indent$}  {:#x}: {}:{} ({})",
                    "",
                    line.address,
                    line.file.name_str(),
                    line.line,
                    line.file.dir_str(),
                    indent = self.1 * 2
                )?;
            }

            write!(f, "{:?}", FunctionsDebug(&function.inlinees, self.1 + 1))?;
        }

        Ok(())
    }
}

#[test]
fn test_breakpad() -> Result<(), Error> {
    // Using the windows version here since it contains all record kinds
    let view = ByteView::open("../testutils/fixtures/windows/crash.sym")?;
    let object = Object::parse(&view)?;

    insta::assert_debug_snapshot_matches!(object, @r###"Breakpad(
    BreakpadObject {
        code_id: Some(
            CodeId(5ab380779000)
        ),
        debug_id: DebugId {
            uuid: "3249d99d-0c40-4931-8610-f4e4fb0b6936",
            appendix: 1
        },
        arch: X86,
        name: "crash.pdb",
        has_symbols: true,
        has_debug_info: true,
        has_unwind_info: true
    }
)"###);

    Ok(())
}

#[test]
fn test_breakpad_symbols() -> Result<(), Error> {
    let view = ByteView::open("../testutils/fixtures/windows/crash.sym")?;
    let object = Object::parse(&view)?;

    let symbols = object.symbol_map();
    insta::assert_debug_snapshot_matches!("breakpad_symbols", SymbolsDebug(&symbols));

    Ok(())
}

#[test]
fn test_breakpad_functions() -> Result<(), Error> {
    let view = ByteView::open("../testutils/fixtures/windows/crash.sym")?;
    let object = Object::parse(&view)?;

    let mut session = object.debug_session()?;
    let functions = session.functions().collect::<Result<Vec<_>, _>>()?;
    insta::assert_debug_snapshot_matches!(
        "breakpad_functions",
        FunctionsDebug(&functions[..10], 0)
    );

    Ok(())
}

#[test]
fn test_elf_executable() -> Result<(), Error> {
    let view = ByteView::open("../testutils/fixtures/linux/crash")?;
    let object = Object::parse(&view)?;

    insta::assert_debug_snapshot_matches!(object, @r###"Elf(
    ElfObject {
        code_id: Some(
            CodeId(f1c3bcc0279865fe3058404b2831d9e64135386c)
        ),
        debug_id: DebugId {
            uuid: "c0bcc3f1-9827-fe65-3058-404b2831d9e6",
            appendix: 0
        },
        arch: Amd64,
        kind: Executable,
        load_address: 0x400000,
        has_symbols: true,
        has_debug_info: false,
        has_unwind_info: true
    }
)"###);

    Ok(())
}

#[test]
fn test_elf_debug() -> Result<(), Error> {
    let view = ByteView::open("../testutils/fixtures/linux/crash.debug")?;
    let object = Object::parse(&view)?;

    insta::assert_debug_snapshot_matches!(object, @r###"Elf(
    ElfObject {
        code_id: Some(
            CodeId(f1c3bcc0279865fe3058404b2831d9e64135386c)
        ),
        debug_id: DebugId {
            uuid: "c0bcc3f1-9827-fe65-3058-404b2831d9e6",
            appendix: 0
        },
        arch: Amd64,
        kind: Debug,
        load_address: 0x400000,
        has_symbols: true,
        has_debug_info: true,
        has_unwind_info: false
    }
)"###);

    Ok(())
}

#[test]
fn test_elf_symbols() -> Result<(), Error> {
    // TODO(ja): Why does crash.debug not retain the symbol table but report has_symbols
    let view = ByteView::open("../testutils/fixtures/linux/crash")?;
    let object = Object::parse(&view)?;

    let symbols = object.symbol_map();
    insta::assert_debug_snapshot_matches!("elf_symbols", SymbolsDebug(&symbols));

    Ok(())
}

#[test]
fn test_elf_functions() -> Result<(), Error> {
    let view = ByteView::open("../testutils/fixtures/linux/crash.debug")?;
    let object = Object::parse(&view)?;

    let mut session = object.debug_session()?;
    let functions = session.functions().collect::<Result<Vec<_>, _>>()?;
    insta::assert_debug_snapshot_matches!("elf_functions", FunctionsDebug(&functions[..10], 0));

    Ok(())
}

#[test]
fn test_mach_executable() -> Result<(), Error> {
    let view = ByteView::open("../testutils/fixtures/macos/crash")?;
    let object = Object::parse(&view)?;

    insta::assert_debug_snapshot_matches!(object, @r###"MachO(
    MachObject {
        code_id: Some(
            CodeId(67e9247c814e392ba027dbde6748fcbf)
        ),
        debug_id: DebugId {
            uuid: "67e9247c-814e-392b-a027-dbde6748fcbf",
            appendix: 0
        },
        arch: Amd64,
        kind: Executable,
        load_address: 0x100000000,
        has_symbols: true,
        has_debug_info: false,
        has_unwind_info: true
    }
)"###);

    Ok(())
}

#[test]
fn test_mach_dsym() -> Result<(), Error> {
    let view =
        ByteView::open("../testutils/fixtures/macos/crash.dSYM/Contents/Resources/DWARF/crash")?;
    let object = Object::parse(&view)?;

    insta::assert_debug_snapshot_matches!(object, @r###"MachO(
    MachObject {
        code_id: Some(
            CodeId(67e9247c814e392ba027dbde6748fcbf)
        ),
        debug_id: DebugId {
            uuid: "67e9247c-814e-392b-a027-dbde6748fcbf",
            appendix: 0
        },
        arch: Amd64,
        kind: Debug,
        load_address: 0x100000000,
        has_symbols: true,
        has_debug_info: true,
        has_unwind_info: false
    }
)"###);

    Ok(())
}

#[test]
fn test_mach_symbols() -> Result<(), Error> {
    let view = ByteView::open("../testutils/fixtures/macos/crash")?;
    let object = Object::parse(&view)?;

    let symbols = object.symbol_map();
    insta::assert_debug_snapshot_matches!("mach_symbols", SymbolsDebug(&symbols));

    Ok(())
}

#[test]
fn test_mach_functions() -> Result<(), Error> {
    let view =
        ByteView::open("../testutils/fixtures/macos/crash.dSYM/Contents/Resources/DWARF/crash")?;
    let object = Object::parse(&view)?;

    let mut session = object.debug_session()?;
    let functions = session.functions().collect::<Result<Vec<_>, _>>()?;
    insta::assert_debug_snapshot_matches!("mach_functions", FunctionsDebug(&functions[..10], 0));

    Ok(())
}

#[test]
fn test_pe_32() -> Result<(), Error> {
    let view = ByteView::open("../testutils/fixtures/windows/crash.exe")?;
    let object = Object::parse(&view)?;

    insta::assert_debug_snapshot_matches!(object, @r###"Pe(
    PeObject {
        code_id: Some(
            CodeId(5ab380779000)
        ),
        debug_id: DebugId {
            uuid: "3249d99d-0c40-4931-8610-f4e4fb0b6936",
            appendix: 1
        },
        debug_file_name: Some(
            "C:\\projects\\breakpad-tools\\windows\\Release\\crash.pdb"
        ),
        arch: X86,
        kind: Executable,
        load_address: 0x400000,
        has_symbols: false,
        has_debug_info: false,
        has_unwind_info: false
    }
)"###);

    Ok(())
}

#[test]
fn test_pe_64() -> Result<(), Error> {
    let view = ByteView::open("../testutils/fixtures/windows/CrashWithException.exe")?;
    let object = Object::parse(&view)?;

    insta::assert_debug_snapshot_matches!(object, @r###"Pe(
    PeObject {
        code_id: Some(
            CodeId(5c9e09599000)
        ),
        debug_id: DebugId {
            uuid: "f535c5fb-2ae8-4bb8-aa20-6c30be566c5a",
            appendix: 1
        },
        debug_file_name: Some(
            "C:\\Users\\sentry\\source\\repos\\CrashWithException\\x64\\Release\\CrashWithException.pdb"
        ),
        arch: Amd64,
        kind: Executable,
        load_address: 0x140000000,
        has_symbols: false,
        has_debug_info: false,
        has_unwind_info: true
    }
)"###);

    Ok(())
}

// NB: No test for PE symbols because our executable does not export any symbols
// NB: No test for PE functions because we can only read debug info from PDBs

#[test]
fn test_pdb() -> Result<(), Error> {
    let view = ByteView::open("../testutils/fixtures/windows/crash.pdb")?;
    let object = Object::parse(&view)?;

    insta::assert_debug_snapshot_matches!(object, @r###"Pdb(
    PdbObject {
        debug_id: DebugId {
            uuid: "3249d99d-0c40-4931-8610-f4e4fb0b6936",
            appendix: 1
        },
        arch: X86,
        load_address: 0x0,
        has_symbols: true,
        has_debug_info: true,
        has_unwind_info: true
    }
)"###);

    Ok(())
}

#[test]
fn test_pdb_symbols() -> Result<(), Error> {
    let view = ByteView::open("../testutils/fixtures/windows/crash.pdb")?;
    let object = Object::parse(&view)?;

    let symbols = object.symbol_map();
    insta::assert_debug_snapshot_matches!("pdb_symbols", SymbolsDebug(&symbols));

    Ok(())
}

#[test]
fn test_pdb_functions() -> Result<(), Error> {
    let view = ByteView::open("../testutils/fixtures/windows/crash.pdb")?;
    let object = Object::parse(&view)?;

    let mut session = object.debug_session()?;
    let functions = session.functions().collect::<Result<Vec<_>, _>>()?;
    insta::assert_debug_snapshot_matches!("pdb_functions", FunctionsDebug(&functions[..10], 0));

    Ok(())
}
