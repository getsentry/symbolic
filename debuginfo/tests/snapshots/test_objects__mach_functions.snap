---
created: "2019-02-20T15:31:37.497922Z"
creator: insta@0.6.3
source: debuginfo/tests/test_objects.rs
expression: "FunctionsDebug(&functions[..10], 0)"
---

> 0xd20: _ZN15google_breakpad18MinidumpFileWriterC2Ev (0x1a)
  0xd20: minidump_file_writer.cc:93 (../deps/breakpad/src/client)
  0xd26: minidump_file_writer.cc:94 (../deps/breakpad/src/client)
  0xd2a: minidump_file_writer.cc:95 (../deps/breakpad/src/client)
  0xd31: minidump_file_writer.cc:96 (../deps/breakpad/src/client)
  0xd39: minidump_file_writer.cc:97 (../deps/breakpad/src/client)

> 0xd40: _ZN15google_breakpad18MinidumpFileWriterC1Ev (0x1a)
  0xd40: minidump_file_writer.cc:96 (../deps/breakpad/src/client)
  0xd46: minidump_file_writer.cc:94 (../deps/breakpad/src/client)
  0xd4a: minidump_file_writer.cc:95 (../deps/breakpad/src/client)
  0xd51: minidump_file_writer.cc:96 (../deps/breakpad/src/client)
  0xd59: minidump_file_writer.cc:97 (../deps/breakpad/src/client)

  > 0xd40: _ZN15google_breakpad18MinidumpFileWriterC2Ev (0x19)
    0xd40: minidump_file_writer.cc:93 (../deps/breakpad/src/client)
    0xd46: minidump_file_writer.cc:94 (../deps/breakpad/src/client)
    0xd4a: minidump_file_writer.cc:95 (../deps/breakpad/src/client)
    0xd51: minidump_file_writer.cc:96 (../deps/breakpad/src/client)

> 0xd60: _ZN15google_breakpad18MinidumpFileWriterD2Ev (0x34)
  0xd60: minidump_file_writer.cc:99 (../deps/breakpad/src/client)
  0xd64: minidump_file_writer.cc:100 (../deps/breakpad/src/client)
  0xd6a: minidump_file_writer.cc:101 (../deps/breakpad/src/client)
  0xd71: minidump_file_writer.cc:133 (../deps/breakpad/src/client)
  0xd7d: minidump_file_writer.cc:140 (../deps/breakpad/src/client)
  0xd84: minidump_file_writer.cc:142 (../deps/breakpad/src/client)
  0xd8a: minidump_file_writer.cc:102 (../deps/breakpad/src/client)
  0xd8c: minidump_file_writer.cc:101 (../deps/breakpad/src/client)

  > 0xd6a: _ZN15google_breakpad18MinidumpFileWriter5CloseEv (0x20)
    0xd6a: minidump_file_writer.cc:127 (../deps/breakpad/src/client)
    0xd71: minidump_file_writer.cc:133 (../deps/breakpad/src/client)
    0xd7d: minidump_file_writer.cc:140 (../deps/breakpad/src/client)
    0xd84: minidump_file_writer.cc:142 (../deps/breakpad/src/client)

> 0xda0: _ZN15google_breakpad18MinidumpFileWriter5CloseEv (0x31)
  0xda0: minidump_file_writer.cc:124 (../deps/breakpad/src/client)
  0xda4: minidump_file_writer.cc:127 (../deps/breakpad/src/client)
  0xdad: minidump_file_writer.cc:133 (../deps/breakpad/src/client)
  0xdbd: minidump_file_writer.cc:140 (../deps/breakpad/src/client)
  0xdc9: minidump_file_writer.cc:142 (../deps/breakpad/src/client)
  0xdcf: minidump_file_writer.cc:146 (../deps/breakpad/src/client)

> 0xde0: _ZN15google_breakpad18MinidumpFileWriterD1Ev (0x34)
  0xde0: minidump_file_writer.cc:99 (../deps/breakpad/src/client)
  0xde4: minidump_file_writer.cc:99 (../deps/breakpad/src/client)
  0xdea: minidump_file_writer.cc:127 (../deps/breakpad/src/client)
  0xdf1: minidump_file_writer.cc:133 (../deps/breakpad/src/client)
  0xdfd: minidump_file_writer.cc:140 (../deps/breakpad/src/client)
  0xe04: minidump_file_writer.cc:142 (../deps/breakpad/src/client)
  0xe0a: minidump_file_writer.cc:102 (../deps/breakpad/src/client)
  0xe0c: minidump_file_writer.cc:101 (../deps/breakpad/src/client)

  > 0xde4: _ZN15google_breakpad18MinidumpFileWriterD2Ev (0x30)
    0xde4: minidump_file_writer.cc:100 (../deps/breakpad/src/client)
    0xdea: minidump_file_writer.cc:101 (../deps/breakpad/src/client)
    0xdf1: minidump_file_writer.cc:133 (../deps/breakpad/src/client)
    0xdfd: minidump_file_writer.cc:140 (../deps/breakpad/src/client)
    0xe04: minidump_file_writer.cc:142 (../deps/breakpad/src/client)
    0xe0c: minidump_file_writer.cc:101 (../deps/breakpad/src/client)

    > 0xdea: _ZN15google_breakpad18MinidumpFileWriter5CloseEv (0x20)
      0xdea: minidump_file_writer.cc:127 (../deps/breakpad/src/client)
      0xdf1: minidump_file_writer.cc:133 (../deps/breakpad/src/client)
      0xdfd: minidump_file_writer.cc:140 (../deps/breakpad/src/client)
      0xe04: minidump_file_writer.cc:142 (../deps/breakpad/src/client)

> 0xe20: _ZN15google_breakpad18MinidumpFileWriter4OpenEPKc (0x49)
  0xe20: minidump_file_writer.cc:104 (../deps/breakpad/src/client)
  0xe27: minidump_file_writer.cc:105 (../deps/breakpad/src/client)
  0xe2c: minidump_file_writer.cc:109 (../deps/breakpad/src/client)
  0xe42: minidump_file_writer.cc:112 (../deps/breakpad/src/client)
  0xe4a: minidump_file_writer.cc:105 (../deps/breakpad/src/client)

> 0xe70: _ZN15google_breakpad18MinidumpFileWriter7SetFileEi (0x2d)
  0xe70: minidump_file_writer.cc:116 (../deps/breakpad/src/client)
  0xe76: minidump_file_writer.cc:117 (../deps/breakpad/src/client)
  0xe78: minidump_file_writer.cc:118 (../deps/breakpad/src/client)
  0xe7c: minidump_file_writer.cc:122 (../deps/breakpad/src/client)
  0xe7e: minidump_file_writer.cc:116 (../deps/breakpad/src/client)

> 0xea0: _ZN15google_breakpad18MinidumpFileWriter20CopyStringToMDStringEPKwjPNS_10TypedMDRVAI8MDStringEE (0x147)
  0xea0: minidump_file_writer.cc:150 (../deps/breakpad/src/client)
  0xeb6: minidump_file_writer.cc:160 (../deps/breakpad/src/client)
  0xed0: minidump_file_writer.cc:161 (../deps/breakpad/src/client)
  0xeda: minidump_file_writer.cc:162 (../deps/breakpad/src/client)
  0xeea: minidump_file_writer.cc:172 (../deps/breakpad/src/client)
  0xefa: minidump_file_writer.cc:174 (../deps/breakpad/src/client)
  0xf0e: minidump_file_writer-inl.h:84 (../deps/breakpad/src/client)
  0xf12: minidump_file_writer.cc:313 (../deps/breakpad/src/client)
  0xf1f: minidump_file_writer.cc:173 (../deps/breakpad/src/client)
  0xf23: minidump_file_writer-inl.h:86 (../deps/breakpad/src/client)
  0xf2a: minidump_file_writer-inl.h:85 (../deps/breakpad/src/client)
  0xf2f: minidump_file_writer-inl.h:86 (../deps/breakpad/src/client)
  0xf34: minidump_file_writer.cc:316 (../deps/breakpad/src/client)
  0xf41: minidump_file_writer.cc:327 (../deps/breakpad/src/client)
  0xf50: minidump_file_writer.cc:328 (../deps/breakpad/src/client)
  0xf6b: minidump_file_writer.cc:166 (../deps/breakpad/src/client)
  0xf72: minidump_file_writer.cc:167 (../deps/breakpad/src/client)
  0xf76: minidump_file_writer.cc:175 (../deps/breakpad/src/client)
  0xf7b: minidump_file_writer.cc:160 (../deps/breakpad/src/client)
  0xf97: minidump_file_writer.cc:179 (../deps/breakpad/src/client)
  0xfa9: minidump_file_writer-inl.h:83 (../deps/breakpad/src/client)
  0xfc8: minidump_file_writer.cc:313 (../deps/breakpad/src/client)

  > 0xefa: _ZN15google_breakpad10TypedMDRVAI8MDStringE20CopyIndexAfterObjectEjPKvm (0xed)
    0xefa: minidump_file_writer-inl.h:83 (../deps/breakpad/src/client)
    0xf0e: minidump_file_writer-inl.h:84 (../deps/breakpad/src/client)
    0xf12: minidump_file_writer-inl.h:84 (../deps/breakpad/src/client)
    0xf23: minidump_file_writer-inl.h:86 (../deps/breakpad/src/client)
    0xf2a: minidump_file_writer-inl.h:85 (../deps/breakpad/src/client)
    0xf2f: minidump_file_writer-inl.h:86 (../deps/breakpad/src/client)
    0xf34: minidump_file_writer.cc:316 (../deps/breakpad/src/client)
    0xf41: minidump_file_writer.cc:327 (../deps/breakpad/src/client)
    0xf50: minidump_file_writer.cc:328 (../deps/breakpad/src/client)
    0xfa9: minidump_file_writer-inl.h:83 (../deps/breakpad/src/client)
    0xfc8: minidump_file_writer.cc:313 (../deps/breakpad/src/client)

    > 0xf12: _ZN15google_breakpad18MinidumpFileWriter4CopyEjPKvl (0xd5)
      0xf12: minidump_file_writer.cc:313 (../deps/breakpad/src/client)
      0xf34: minidump_file_writer.cc:316 (../deps/breakpad/src/client)
      0xf41: minidump_file_writer.cc:327 (../deps/breakpad/src/client)
      0xf50: minidump_file_writer.cc:328 (../deps/breakpad/src/client)
      0xfc8: minidump_file_writer.cc:313 (../deps/breakpad/src/client)

> 0xff0: _ZN15google_breakpad18MinidumpFileWriter20CopyStringToMDStringEPKcjPNS_10TypedMDRVAI8MDStringEE (0x147)
  0xff0: minidump_file_writer.cc:183 (../deps/breakpad/src/client)
  0x1006: minidump_file_writer.cc:189 (../deps/breakpad/src/client)
  0x1020: minidump_file_writer.cc:190 (../deps/breakpad/src/client)
  0x1030: minidump_file_writer.cc:191 (../deps/breakpad/src/client)
  0x1039: minidump_file_writer.cc:199 (../deps/breakpad/src/client)
  0x1049: minidump_file_writer.cc:201 (../deps/breakpad/src/client)
  0x1058: minidump_file_writer-inl.h:84 (../deps/breakpad/src/client)
  0x105b: minidump_file_writer.cc:313 (../deps/breakpad/src/client)
  0x1067: minidump_file_writer.cc:200 (../deps/breakpad/src/client)
  0x106b: minidump_file_writer-inl.h:86 (../deps/breakpad/src/client)
  0x1072: minidump_file_writer-inl.h:85 (../deps/breakpad/src/client)
  0x1076: minidump_file_writer-inl.h:86 (../deps/breakpad/src/client)
  0x107b: minidump_file_writer.cc:316 (../deps/breakpad/src/client)
  0x108c: minidump_file_writer.cc:327 (../deps/breakpad/src/client)
  0x109b: minidump_file_writer.cc:328 (../deps/breakpad/src/client)
  0x10b7: minidump_file_writer.cc:195 (../deps/breakpad/src/client)
  0x10ba: minidump_file_writer.cc:196 (../deps/breakpad/src/client)
  0x10c0: minidump_file_writer.cc:202 (../deps/breakpad/src/client)
  0x10c5: minidump_file_writer.cc:189 (../deps/breakpad/src/client)
  0x10e7: minidump_file_writer.cc:205 (../deps/breakpad/src/client)
  0x10f9: minidump_file_writer-inl.h:83 (../deps/breakpad/src/client)
  0x1118: minidump_file_writer.cc:313 (../deps/breakpad/src/client)

  > 0x1049: _ZN15google_breakpad10TypedMDRVAI8MDStringE20CopyIndexAfterObjectEjPKvm (0xee)
    0x1049: minidump_file_writer-inl.h:83 (../deps/breakpad/src/client)
    0x1058: minidump_file_writer-inl.h:84 (../deps/breakpad/src/client)
    0x105b: minidump_file_writer-inl.h:84 (../deps/breakpad/src/client)
    0x106b: minidump_file_writer-inl.h:86 (../deps/breakpad/src/client)
    0x1072: minidump_file_writer-inl.h:85 (../deps/breakpad/src/client)
    0x1076: minidump_file_writer-inl.h:86 (../deps/breakpad/src/client)
    0x107b: minidump_file_writer.cc:316 (../deps/breakpad/src/client)
    0x108c: minidump_file_writer.cc:327 (../deps/breakpad/src/client)
    0x109b: minidump_file_writer.cc:328 (../deps/breakpad/src/client)
    0x10f9: minidump_file_writer-inl.h:83 (../deps/breakpad/src/client)
    0x1118: minidump_file_writer.cc:313 (../deps/breakpad/src/client)

    > 0x105b: _ZN15google_breakpad18MinidumpFileWriter4CopyEjPKvl (0xdc)
      0x105b: minidump_file_writer.cc:313 (../deps/breakpad/src/client)
      0x107b: minidump_file_writer.cc:316 (../deps/breakpad/src/client)
      0x108c: minidump_file_writer.cc:327 (../deps/breakpad/src/client)
      0x109b: minidump_file_writer.cc:328 (../deps/breakpad/src/client)
      0x1118: minidump_file_writer.cc:313 (../deps/breakpad/src/client)

> 0x1140: _ZN15google_breakpad18MinidumpFileWriter11WriteStringEPKwjP20MDLocationDescriptor (0x5)
  0x1140: minidump_file_writer.cc:245 (../deps/breakpad/src/client)

